File: CompilationEngine.js
fs = require('fs');
buffer = [];
tokens = [];
output = [];
var token;
var previousToken;

op = ["+", "-", "*", "/", "&", "|", "<", ">", "="];
keyword_constant = ["true", "false", "null", "this"];
unaryop = ["-", "~"];

function debug() {
    console.log("*** DEBUG ***");
    console.log("previous token: " + previousToken.token);
    console.log("current token: " + token.token);
    console.log("next token: " + tokens[0].token);
};

function write(cmd) {
	output.push(cmd);
};

function writeOpen(cmd) {
    output.push("<" + cmd + ">")
}

function writeClose(cmd) {
    output.push("</" + cmd + ">")
}

function writeToken() {
    write(["<" + token.token_type + ">" + " " + token.token + " " + "</" + token.token_type + ">" ]);    
};

function genOutFile() {
	output = output.join("\n");
	path = process.argv[2].split("/")
    name = process.argv[3]
	fd = fs.openSync(path[0] + "/" + path[1] + "/" + name + "_.xml", 'w');
	fs.write(fd, output);
};

function hasMoreTokens() {
    if(tokens.length) {
        return true;
    } else {
        return false;
    };
};

function advance() {
    previousToken = token;
    token = tokens.shift();
};

function generateArray(buffer) {
    for (var i=0; i<buffer.length; i++) {
        temp = buffer[i].split(/"/).filter(function(e){return e!="" && e!=null && e!=" "});
        tokens.push({"token_type":temp[0], "token":temp[1]});
    }
};

function raiseError(message) {
    console.error("error: " + message);
};

function checkToken() {
    var args = Array.prototype.slice.call(arguments);
    if (args.indexOf(token.token) == -1 ) { return false };
    return true;
};

function checkTokenType() {
    var args = Array.prototype.slice.call(arguments);
    if (args.indexOf(token.token_type) == -1 ) { return false };
    return true;
};

function checkNextToken() {
    var args = Array.prototype.slice.call(arguments);
    if (args.indexOf(tokens[0].token) == -1) { return false };
    return true;
};

function checkIdentifier() {
    if ((token.token).match(/^[a-zA-Z_][\w]*$/)) { 
        return true;
    } else {
        raiseError("invalid identifier");
        return false;
    }
};

function checkSemicolon() {
    if (checkToken(";")) {
        writeToken();
        advance();
        return true;
    } else {
        raiseError("missing end of line semicolon");
        return false;
    }   
};

function checkTypeAndIdentifier() {
    //check type
    if (!checkToken("int", "char", "boolean") && !checkIdentifier()) {
        raiseError("missing type specifier in var dec");
        return false;
    }
    
    writeToken();
    advance();

    //check variable name
    if (!checkIdentifier) { raiseError("var dec identifier is invalid") ; return false };
    
    writeToken();
    advance();
    return true;
};

function checkOpeningBracket() {
    if (checkToken("(")) {
        writeToken();
        advance();
        return true;
    } else {
        raiseError("missing opening bracket");
        return false;
    }
};

function checkClosingBracket() {
    if (checkToken(")")) {
        writeToken();
        advance();
        return true;
    } else {
        raiseError("missing closing bracket");
        return false;
    }
};

function checkOpeningBrace() {
    if (checkToken("{")) {
        writeToken();
        advance();
        return true;
    } else {
        raiseError("missing opening brace");
        return false;
    }
};

function checkClosingBrace() {
    if (checkToken("}")) {
        writeToken();
        advance();
        return true;
    } else {
        raiseError("missing closing brace");
        return false;
    }
};

function checkSubroutineCall() {
    
    if (!checkIdentifier()) { return false };
    
    writeToken();
    advance();

    //could be subroutineName(blargh) | class/var.subroutine.(blargh)

    if (checkToken(".")) {
        writeToken();
        advance();
        
        if (!checkIdentifier()) { return false };
        
        writeToken();
        advance();
    }
    
    if (!checkOpeningBracket()) { return false };
    if (!compileExpressionList()) { return false };
    if (!checkClosingBracket()) { return false };

    return true;
};

function compileTerm() {
    
    writeOpen("term")
    ///int, string or keyword
    if (checkTokenType("integerConstant", "stringConstant", "keywordConstant")) { 
        writeToken();
        advance();
        return true;
    }

    //(expression)
    if (checkToken("(")) {
        if (!compileExpression()) { raiseError("compileExpression") ; return false };
        if (!checkClosingBracket()) { return false }
        return true;
    }

    //unaryOp term    
    if (unaryop.indexOf(token.token) != -1) {
        writeToken();
        advance();

        if (!compileTerm()) { return false };
        return true;
    }

    //can now only have varName | varName [expression] | subroutineCall
    //All of these terms begin with a varName ==> look ahead one token to differentiate


    if (checkNextToken("[")) {
        if (!checkIdentifier()) { return false };
        writeToken();
        advance();

        //write "["
        writeToken();
        advance();

        if (!compileExpression()) { return false };
        if (!checkToken("]")) { return false };

        writeToken();
        advance();

        return true;
    }

    if (checkNextToken("(")) {
        if (!checkSubroutineCall()) { return false };
        return true;
    }

    if (!checkIdentifier()) { return false } 
    writeToken();
    advance();
    writeClose("term")
    return true;
};

function compileClass() {
    advance();
    if (checkToken("class")) {
        writeOpen("class")
        writeToken();
        advance();
        if (!checkIdentifier()) { raiseError("invalid class identifier"); return false };   
        writeToken();
        advance();
        if (!checkToken("{")) {
            raiseError("no opening brace for class");
            return false;
        }
        writeToken();
        advance();
        
        //classVarDec
        while (checkToken("static") || checkToken("field")) {
            if (!compileClassVarDec()) { return false };
        }

        //subroutineDec
        while (checkToken("constructor") || checkToken("function") || checkToken("method")) {
            if (!compileSubroutine()) { raiseError("compileSubroutine"); return false };
        }

        if (!checkClosingBrace()) { return false }

        if (tokens.length != 0) {
            raiseError("tokens present after class closed");
            return false;
        }
        writeClose("class")
        return true;

    } else {
        raiseError("missing class keyword to open");
        return false;
    }
};

function compileClassVarDec() {
    writeOpen("classVarDec")
    writeToken();
    advance();
    //check type
    if (!checkToken("int", "char", "bool") && !checkIdentifier()) {
        raiseError("missing type specifier");
        return false;
    }
    writeToken();
    advance();
    if (!checkIdentifier()) { raiseError("checkIdentifier") ; return false };
    writeToken();
    advance();
    
    //deal with (',', varName)*
    while (token.token == ",") {
        writeToken();
        advance();
        if (!checkIdentifier()) { return false };
        writeToken();
        advance(); 
    }

    if (!checkSemicolon()) { return false };
    
    writeClose("classVarDec")
    return true;
}

function compileSubroutine() {
    writeOpen("subroutineDec")
    writeToken();
    advance();

    //check return type
    if (!checkToken("void", "int", "char", "bool")  && !checkIdentifier()) {
        raiseError("missing subroutine return type");
        return false;
    }
    
    writeToken();
    advance();

    //check subroutine name
    if (!checkIdentifier()) { raiseError("invalid subroutine identifier") ; return false };
    writeToken();
    advance();

    //opening bracket
    if (!checkOpeningBracket()) { return false };

    //parameter list
    if (!compileParameterList()) { return false };
    
    //closing bracket
    if (!checkClosingBracket()) { return false };

    writeOpen("subroutineBody");

    //opening brace
    if (!checkOpeningBrace()) { return false };
    
    //possible varDec
    while (checkToken("var")) {  
        if (!compileVarDec()) { 
            raiseError("error: compileVarDec");
            return false;
        }
    }

    //compile statements
    if (!compileStatements()) { 
        raiseError("compileStatements");
        return false;
    }

    //closing brace
    if (!checkClosingBrace()) { return false };

    writeClose("subroutineBody");

    writeClose("subroutineDec");
    return true;
};

function compileParameterList() {
    //no paramters
    writeOpen("parameterList")
    if (checkToken(")")) { writeClose("parameterList") ; return true };

    if (!checkTypeAndIdentifier()) { raiseError("checkTypeAndIdentifier") ; return false };

    while(checkToken(",")) {
        if (!checkTypeAndIdentifier()) { return false };
    }
    
    writeClose("parameterList")
    return true;
};

function compileVarDec() {
    writeOpen("varDec")
    writeToken();
    advance();

    if (!checkTypeAndIdentifier()) { raiseError("checkTypeAndIdentifier") ; return false };

    while (checkToken(",")) {
        if (!checkIdentifier()) { return false };
        writeToken();
        advance();
    }

    if (!checkSemicolon()) { return false };
    
    writeClose("varDec")
    return true;
};

function compileStatements() {
    writeOpen("statements")
    while (checkToken("let", "if", "while", "do", "return")) {

        switch(token.token) {
            case "let":
                if (!compileLet()) { raiseError("compileLet") ; return false };
                break;
            case "if":
                if (!compileIf()) { raiseError("compileIf") ; return false };
                break;
            case "while":
                if (!compileWhile()) { raiseError("compileWhile") ; return false };
                break;
            case "do":
                if (!compileDo()) { raiseError("compileDo") ; return false };
                break;
            case "return":
                if (!compileReturn()) { raiseError("compileReturn") ; return false };
                break;
            default:
                raiseError("invalid statement keyword");
                return false;
        }
    }
    writeClose("statements")
    return true;
};

function compileLet() {
    writeOpen("letStatement")
    writeToken();
    advance();

    //varName 
    if (!checkIdentifier()) { raiseError("invalid compileLet identifier") ; return false };

    writeToken();
    advance();

    //potential expression
    if (checkToken("[")) {
        writeToken();
        advance();
    
        compileExpression();

        if (!checkToken("]")) { return false };
        
        writeToken();
        advance();
    }

    //equals expression
    if (!checkToken("=")) { raiseError("missing =") ; return false };

    writeToken();
    advance();

    if (!compileExpression()) { raiseError("unable to compileExpression inside compileLet") ; return false }

    if (!checkSemicolon()) { return false }; 
    
    writeClose("letStatement");
    return true;
};

function compileIf() {
    writeOpen("ifStatment");
    writeToken();
    advance();

    //varName
    if (!checkIdentifier()) { return false };

    writeToken();
    advance();

    //expression
    if (!checkOpeningBracket()) { return false };

    writeToken();
    advance();

    if (!compileExpression()) { return false };

    if (!checkClosingBracket()) { return false };

    if (!checkOpeningBrace()) { return false };
    
    if (!compileStatements()) { return false };

    if (!checkClosingBrace()) { return false };

    if (checkToken("else")) {
        if (!checkOpeningBrace()) { return false };
        if (!compileStatements()) { return false };
        if (!checkClosingBrace()) { return false };
    }
    writeClose("ifStatement");
    return true
};

function compileWhile() {
    writeOpen("whileStatement");
    writeToken();
    advance();

    if (!checkOpeningBracket()) { return false };

    if (!compileExpression()) { return false };
    
    if (!checkClosingBracket()) { return false };

    if (!checkOpeningBrace()) { return false };

    if (!compileStatements()) { return false };

    if (!checkClosingBrace()) { return false };
    
    writeClose("whileStatement");
    return true;
};

function compileDo() {
    writeOpen("doStatement");
    writeToken();
    advance();

    if (!checkSubroutineCall()) { return false };
    if (!checkSemicolon()) { return false };
    writeClose("doStatement");
    return true;
};

function compileReturn() {
    writeOpen("returnStatement");
    writeToken();
    advance();
    //expression is optional
    if (checkToken(";")) {
        writeToken();
        advance();
        writeClose("returnStatement")
        return true;
    }

    if (!compileExpression()) { raiseError("compileExpression") ; return false };
    if (!checkSemicolon()) { return false };
    writeClose("returnStatement");
    return true;
};

function compileExpression() {
    writeOpen("expression");
    //an expression must contain at least one term
    if (!compileTerm()) { raiseError("compileTerm") ; return false };

    //(op term)*
    while (op.indexOf(token.token) != -1) {
        writeToken();
        advance();
        if (!compileTerm()) { return false };
    }
    writeClose("expression");
    return true;
};

function compileExpressionList() {
    writeOpen("expressionList");
    //first off, check if we have an empty list
    if (checkToken(")")) { writeClose("expressionList") ; return true };
    
    //now we know we have to compile at least one expression
    if (!compileExpression()) { return false };

    while (!checkToken(")")) { 
        if (!checkToken(",")) { return false };
        writeToken();
        advance();
        if (!compileExpression()) { return false };
    }
    writeClose("expressionList");
    return true;
};

function main() {
    file = process.argv[2] 
    data = fs.readFileSync(file, 'ascii');
    buffer = data.split(/[\n]/);
    generateArray(buffer);
    
    //first routine to be called must be compileClass
    if (!compileClass()) { raiseError("unable to compile class") ; debug() ; return }
	genOutFile();
};

main();

File: src/10/CompilationEngine.go
package main

import (
    "os"
    "fmt"
    "io/ioutil"
    "strings"
    "regexp"
    "sort"
)

var tokens [][]string
var current []string
var output []string

var operators = []string{"+", "-", "*", "/", "&amp;", "|", "&lt;", "&gt;", "="}
var unaryOp = []string{"-", "~"}

func generateTokenArray(slice []string) {

    for i := range slice {
        pair := strings.Split(slice[i], " ")
        tokens = append(tokens, pair)
    }
    advance()
}

func write(cmd string) {
    output = append(output, cmd)
}

func writeOpen(cmd string) {
    output = append(output, "<" + cmd + ">")
}

func writeClose(cmd string) {
    output = append(output, "</" + cmd + ">")
}

func writeToken() {
    //do this better
    formatted := "<" + current[0] + ">" + " " + current[1] + " " + "</" + current[0] + ">"
    output = append(output, formatted)
    advance()
}

func advance() {
    current = tokens[0]
    tokens = tokens[1:]
}

func raiseError(message string) {
    fmt.Println("error: ", message)
}

func checkToken(token string) bool {
    if strings.Compare(token, current[1]) == 0 {
        return true
    }
    return false
}

func checkTokenSlice(slice []string) bool {
    for i := range slice {
        if strings.Compare(slice[i], current[1]) == 0 {
            return true
        }
    }
    return false
}

func checkTokenTypeSlice(slice []string) bool {
    for i := range slice {
        if strings.Compare(slice[i], current[0]) == 0 {
            return true
        }
    }
    return false
}

func checkIdentifier() bool {
    re, err := regexp.Compile("^[a-zA-Z_][\\w]*$")
    if err != nil {
        os.Exit(1)
    }
    if !re.MatchString(current[1]) {
        raiseError("invalid identifier")
        return false
    }
    writeToken()
    return true
}

func checkIdentifierPassive() bool {
    re, err := regexp.Compile("^[a-zA-Z_][\\w]*$")
    if err != nil {
        os.Exit(1)
    }
    if !re.MatchString(current[1]) {
        return false
    }
    return true
}

func checkOpeningBrace() bool {
    if !checkToken("{") {
        raiseError("missing opening brace")
        return false
    }
    writeToken()
    return true
}

func checkClosingBrace() bool {
    if !checkToken("}") {
        raiseError("missing closing brace")
        return false
    }
    writeToken()
    return true
}

func checkOpeningBracket() bool {
    if !checkToken("(") {
        raiseError("missing opening bracket")
        return false
    }
    writeToken()
    return true
}

func checkClosingBracket() bool {
    if !checkToken(")") {
        raiseError("missing closing bracket")
        return false
    }
    writeToken()
    return true
}

func checkSemicolon() bool {
    if !checkToken(";") {
        raiseError("missing semicolon")
        return false
    }
    writeToken()
    return true
}

func checkTypeAndIdentifier() bool {
    //check type
    if (!checkTokenSlice([]string{"int", "char", "bool"}) && !checkIdentifierPassive()) {
        raiseError("type specifier")
        return false
    }

    writeToken()

    if !checkIdentifier() {
        return false
    }
    return true
}

func checkNextToken(token string) bool {
    if strings.Compare(token, tokens[0][1]) == 0 {
        return true
    }
    return false
}

func debug() {
    fmt.Println("current: ", current[1])
    fmt.Println("next: ", tokens[0][1])

    for i := range output {
        fmt.Println(output[i])
    }
}

func compileClass() bool {

    if !checkToken("class") {
        raiseError("missing class keyword to open")
        return false
    }

    writeOpen("class")
    writeToken()

    if !checkIdentifier() { return false }

    if !checkOpeningBrace() { return false }

    for (checkTokenSlice([]string{"static", "field"})) {
        if !compileClassVarDec() {
            raiseError("compileClassVarDec")
            return false
        }
    }

    for checkTokenSlice([]string{"constructor", "function", "method"}) {
        if !compileSubroutine() {
            raiseError("compileSubroutine")
            return false
        }
    }

    if !checkClosingBrace() { return false }

    if len(tokens) != 0 {
        raiseError("tokens remaining after class closed")
        return false
    }

    writeClose("class")
    return true
}

func compileClassVarDec() bool {
    writeOpen("classVarDec")
    writeToken()

    //check type
    if (!checkTokenSlice([]string{"int, char, bool"}) && !checkIdentifierPassive()) {
        raiseError("missing type specifier")
        return false
    }

    writeToken()

    if !checkIdentifier() { return false }

    //deal with (',', varName)*
    for checkToken(",") {
        writeToken()
        if !checkIdentifier() {
            return false
        }
    }

    if !checkSemicolon() { return false }

    writeClose("classVarDec")
    return true
}

func compileSubroutine() bool {
    writeOpen("subroutineDec")
    writeToken()

    //check return type
    if !checkTokenSlice([]string{"void", "int", "char", "bool"}) && !checkIdentifierPassive() {
        raiseError("missing return type")
        return false
    }

    writeToken()

    //check subroutine name
    if !checkIdentifier() {
        raiseError("invalid identifier")
        return false
    }

    if !checkOpeningBracket() { return false }

    if !compileParameterList() {
        raiseError("compileParameterList")
        return false
    }

    if !checkClosingBracket() { return false }

    writeOpen("subroutineBody")

    if !checkOpeningBrace() { return false }

    //possible variable declaration
    for checkToken("var") {
        if !compileVarDec() {
            raiseError("compileVarDec")
            return false
        }
    }

    //compile statements
    if !compileStatements() {
        raiseError("compileStatements")
        return false
    }

    if !checkClosingBrace() { return false }

    writeClose("subroutineBody")
    writeClose("subroutineDec")

    return true
}

func compileParameterList() bool {
    writeOpen("parameterList")

    //no parameters
    if checkToken(")") {
        writeClose("parameterList")
        return true
    }

    if !checkTypeAndIdentifier() {
        raiseError("checkTypeAndIdentifier")
        return false
    }

    for checkToken(",") {
        writeToken()
        if !checkTypeAndIdentifier() {
            raiseError("checkTypeAndIdentifier")
            return false
        }
    }

    writeClose("parameterList")
    return true
}

func compileVarDec() bool {
    writeOpen("varDec")
    writeToken()

    if !checkTypeAndIdentifier() {
        raiseError("checkTypeAndIdentifier")
        return false
    }

    for checkToken(",") {
        writeToken()
        if !checkIdentifier() { return false }
    }

    if !checkSemicolon() { return false }

    writeClose("varDec")
    return true
}

func compileStatements() bool {
    writeOpen("statements")
    for checkTokenSlice([]string{"let", "if", "while", "do", "return"}) {

        switch current[1] {
            case "let":
                if !compileLet() {
                    raiseError("compileLet")
                    return false
                }
                break
            case "if":
                if !compileIf() {
                    raiseError("compileIf")
                    return false
                }
                break
            case "while":
                if !compileWhile() {
                    raiseError("compileWhile")
                    return false
                }
                break
            case "do":
                if !compileDo() {
                    raiseError("compileDo")
                    return false
                }
                break
            case "return":
                if !compileReturn() {
                    raiseError("compileReturn")
                    return false
                }
                break
            default:
                raiseError("invalid statement keyword")
                return false
        }
    }
    writeClose("statements")
    return true
}

func compileLet() bool {
    writeOpen("compileLet")
    writeToken()

    if !checkIdentifier() { return false }

    //potential expression
    if checkToken("[") {
        writeToken()

        if !compileExpression() {
            raiseError("compileExpression")
            return false
        }

        if !checkToken("]") {
            raiseError("missing closing ]")
            return false
        }
        writeToken()
    }

    //equals expression
    if !checkToken("=") {
        raiseError("missing =")
        return false
    }

    writeToken()

    if !compileExpression() {
        raiseError("compileExpression")
        return false
    }

    if !checkSemicolon() { return false }

    writeClose("letStatement")
    return true
}

func compileIf() bool {
    writeOpen("ifStatement")
    writeToken()

    if !checkOpeningBracket() { return false }
    if !compileExpression() {
        raiseError("compileExpression")
        return false
    }
    if !checkClosingBracket() { return false }
    if !checkOpeningBrace() { return false }
    if !compileStatements() {
        raiseError("compileStatements")
        return false
    }
    if !checkClosingBrace() { return false }
    if checkToken("else") {
        if !checkOpeningBrace() { return false }
        if !compileStatements() { return false }
        if !checkClosingBrace() { return false }
    }
    writeClose("ifStatement")
    return true
}

func compileWhile() bool {
    writeOpen("whileStatement")
    writeToken()

    if !checkOpeningBracket() { return false }
    if !compileExpression() {
        raiseError("compileExpression")
        return false
    }
    if !checkClosingBracket() { return false }
    if !checkOpeningBrace() { return false }
    if !compileStatements() {
        raiseError("compileStatements")
        return false }
    if !checkClosingBrace() { return false }

    writeClose("whileStatement")
    return true
}

func compileDo() bool {
    writeOpen("doStatement")
    writeToken()

    if !checkSubroutineCall() {
        raiseError("subroutineCall")
        return false
    }
    if !checkSemicolon() { return false }

    writeClose("doStatement")
    return true
}

func compileReturn() bool {
    writeOpen("returnStatement")
    writeToken()

    //expression is optional
    if checkToken(";") {
        writeToken()
        writeClose("returnStatement")
        return true
    }

    if !compileExpression() {
        raiseError("compileExpression")
        return true
    }
    if !checkSemicolon() { return false }

    writeClose("returnStatement")
    return true
}

func compileExpression() bool {
    writeOpen("expression")
    //an expression must contain at least one term
    if !compileTerm() {
        raiseError("compileTerm")
        return false
    }

    //(op term)* 
    op := true
    for op {
        for i := range operators {
            if strings.Compare(operators[i], current[1]) == 0 {
                op = true
                writeToken()
                if !compileTerm() {
                    raiseError("compileTerm")
                    return false
                }
                break
            }
        op = false
        }
    }

    writeClose("expression")
    return true
}

func compileTerm() bool {
    writeOpen("term")

    //int, string or keyword
    if checkTokenTypeSlice([]string{"integerConstant", "stringConstant", "keywordConstant"}) {
        writeToken()
        return true
    }

    //(expression)
    if checkToken("(") {
        writeToken()
        if !compileExpression() {
            raiseError("compileExpression")
            return false
        }
        if !checkClosingBracket() {
            return false
        }
        return true
    }

    //unaryOp term
    for i := range unaryOp {
        if strings.Compare(unaryOp[i], current[1]) == 0 {
            writeToken()
            if !compileTerm() {
                raiseError("compileTerm")
                return false
            }
            return true
        }
    }

    //can now only have varName | varName [expression] | subroutineCall
    //Al of these terms begin with varName => look ahead one token to
    //differentiate

    if checkNextToken("[") {
        if !checkIdentifier() { return false }
        writeToken()

        if !compileExpression() { 
            raiseError("compileExpression")
            return false
        }

        if !checkToken("]") {
            raiseError("missing ]")
            return false
        }
        writeToken()

        return true
    }

    if checkNextToken("(") || checkNextToken(".") {
        if !checkSubroutineCall() {
            raiseError("checkSubroutineCall")
            return false
        }
        return true
    }

    if !checkIdentifier() { return false }

    writeClose("term")
    return true
}

func checkSubroutineCall() bool {
    if !checkIdentifier() { return false }

    //could be subroutineName(blargh) | class/var.subroutine(blargh)

    if checkToken(".") {
        writeToken()
        if !checkIdentifier() { return false }
    }

    if !checkOpeningBracket() { return false }

    if !compileExpressionList() { 
        raiseError("compileExpressionList")
        return false
    }
    if !checkClosingBracket() { return false }

    return true
}

func compileExpressionList() bool {
    writeOpen("expressionList")
    //first off, check if we have an empty list
    if checkToken(")") {
        writeClose("expressionList")
        return true
    }

    //now we know we have to compile at least one expression
    if !compileExpression() {
        raiseError("compileExpression")
        return false
    }

    for !checkToken(")") {
        if !checkToken(",") {
            raiseError("missing comma")
            return false
        }
        writeToken()
        if !compileExpression() {
            raiseError("compileExpression")
            return false
        }
    }
    writeClose("expressionList")
    return true
}

func main () {

    //arguments
    args := os.Args
    filepath := args[1]
    target := args[2]

    //open tokens file
    data, err := ioutil.ReadFile(filepath)
    if err != nil {
        os.Exit(1)
    }

    stringified := string(data)
    re, _ := regexp.Compile("\"")
    stringified = re.ReplaceAllString(stringified, "")
    slice := strings.Split(stringified, "\n")
    generateTokenArray(slice)

    //sort op slice
    sort.Strings(operators)

    //first routine to be called must be compileClass
    if !compileClass() {
        raiseError("unable to compile class")
        debug()
    }

    //write output to file
    f, err := os.Create(target)
    if err != nil {
        os.Exit(1)
    }
    defer f.Close()

    for i := range output {
        f.WriteString(output[i])
        f.WriteString("\n")
    }
}
